src/main/java/org/jenkinsci/remoting/nio/FifoBuffer.java

```diff
@@ -250,16 +251,19 @@ public int send(WritableByteChannel ch) throws IOException {
                     }
                     return 0;   // no data to read
                 }
+                try {
+                    int sent = r.send(ch, chunk);  // bytes actually written
 
-                int sent = r.send(ch, chunk);  // bytes actually written
-
-                read += sent;
-                sz -= sent;
+                    read += sent;
+                    sz -= sent;
 
-                lock.notifyAll();
+                    lock.notifyAll();
 
-                if (sent==0)    // channel filled up
-                    return read;
+                    if (sent == 0)    // channel filled up
+                        return read;
+                } catch (ClosedChannelException e) {
+                    return -1; // propagate EOF
```

Comment:
Not in my opinion... as `-1` means we have reached the EOF... doesn't matter how much we sent, we can send no more...

If people have strong views on this I am ok removing the handling from `send`... it's the `receive` side that is annoying me with pointless stack traces